libtenzir/src/table_slice.cpp

```diff
@@ -1302,4 +1301,241 @@ auto unflatten(const table_slice& slice,
+  using variant::variant;
+
+  friend auto inspect(auto& f, unflatten_entry& x) -> bool {
+    if (auto dbg = as_debug_writer(f)) {
+      return x.match(
+        [&](const std::shared_ptr<arrow::Array>& array) {
+          return dbg->fmt_value("<{} array", array->type()->ToString());
+        },
+        [&](const unflatten_record& record) {
+          return dbg->apply(record);
+        });
+    }
+    return false;
+  }
+};
+
+unflatten_record::unflatten_record(int64_t length,
+                                   std::shared_ptr<arrow::Buffer> null_bitmap)
+  : length{length}, null_bitmap{std::move(null_bitmap)} {
+}
+
+} // namespace
+
+// {x.y: int64, x: {z: int64}} -> {x: {y: int64, z: int64}}
+//
+// {x.y: 1, x: {z: 2}} ->  {x: {y: 1, z: 2}}
+// {x.y: null, x: {z: 2}} ->  {x: {y: null, z: 2}}
+// {x.y: 1, x: null} ->  {x: null} ?? {x: {y: 1, z: null}}
+// {x.y: null, x: null} ->  {x: null} ?? {x: {y: null, z: null}}
+//
+// {x: {y: int64}} -> {x: {y: int64}}
+// {x: null} -> {x: null}
+//
+// {x.y: int64, x: {}}
+// {x.y: 1, x: null} -> {x.y: 1}
+// {x.y: null, x:null} -> {x: {y: null}}
+// {x.y: null, x: {}}, {x: {y: null}}
+//
+// {x: {x.y: int64}}
+// {x: null}
+//
+// => If at least one entry is non-null, then result is non-null.
+//    (taking care to propagate nulls down)
+// => If all entries are null, then then the result is null.
+//    (where `{x.y: null}` makes `x` non-null)
+
+// auto unflatten(const arrow::Array& array, std::string_view sep)
+//   -> std::shared_ptr<arrow::Array>;
+
+void unflatten_recurse(unflatten_entry& entry,
+                       std::shared_ptr<arrow::Array> array,
+                       std::string_view sep);
+
+auto unflatten(const arrow::ListArray& array, std::string_view sep)
+  -> std::shared_ptr<arrow::ListArray>;
+
+auto bitmap_or(std::shared_ptr<arrow::Buffer> x,
+               std::shared_ptr<arrow::Buffer> y)
+  -> std::shared_ptr<arrow::Buffer> {
+  if (not x || not y) {
+    return nullptr;
+  }
+  auto size = std::min(x->size(), y->size());
+  auto z = check(arrow::AllocateBuffer(size));
+  // TODO: Is this safe?
+  auto x_ptr = x->data();
+  auto y_ptr = y->data();
+  auto z_ptr = z->mutable_data();
+  for (auto i = int64_t{0}; i < size; ++i) {
+    z_ptr[i] = x_ptr[i] | y_ptr[i]; // NOLINT
+  }
+  return z;
+}
+
+void unflatten_recurse(unflatten_entry& root, const arrow::StructArray& array,
+                       std::string_view sep) {
+  auto ty = array.struct_type();
+  auto names = ty->fields() | std::views::transform(&arrow::Field::name);
+  // We need to flatten the null bitmap here because it can happen that the
+  // fields are saved to a record that is made non-null by another entry.
+  auto fields = check(array.Flatten());
+  if (not std::holds_alternative<unflatten_record>(root)) {
+    root.emplace<unflatten_record>(array.length(), array.null_bitmap());
+  }
+  for (auto [name, data] : detail::zip_equal(names, fields)) {
+    // TODO: Empty name? Double separators?
```

Comment:
The double separators should result in an empty field name, right? I think the old implementation did not handle that correctly.

Reply:
Seems like `std::views::split` does the right thing for double separators, but not for empty name. The empty name should still be the empty name after splitting, but it returns nothing (seems like a bad choice to me). Added a workaround for that.